### PR TITLE
Propagate CUDAOutOfMemoryError to Python.

### PR DESCRIPTION
**Pull request by zdevito**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #83146

The intention is to make it easier to catch this situation for debugging,
logging, or application-specific recovery.

### CODE REVIEW

File: torch/csrc/Exceptions.cpp [hunk 1]
```diff
@@ -51,6 +52,14 @@ could not be completed because the input matrix is singular.",
       PyModule_AddObject(module, "_LinAlgError", THPException_LinAlgError) ==
       0);
 
+  ASSERT_TRUE(
+      THPException_CUDAOutOfMemoryError = PyErr_NewException(
+          "torch.cuda.CUDAOutOfMemoryError", nullptr, nullptr));
```
Comment (albanD) [inline torch/csrc/Exceptions.cpp:57]: Should this inherit from RuntimeError or is the default Exception ok?

**Review by albanD — APPROVED** [review]
Nice cleanup of the Exceptions code. Thanks!

File: torch/csrc/Exceptions.cpp (hunk 1 above)
Comment (zdevito) [reply torch/csrc/Exceptions.cpp:57]: For bc it probably should.

**Comment by zdevito** [comment]
Switching the name of the C++ exception to `c10::OutOfMemoryError` (and `torch.cuda.OutOfMemoryError`) because having it be accessible outside of CUDA confuses the hipify process. This complexity may be why it wasn't exposed before.

File: c10/util/Exception.h [hunk 2]
```diff
@@ -235,6 +235,10 @@ class C10_API LinAlgError : public Error {
   using Error::Error;
 };
 
+class C10_API OutOfMemoryError : public Error {
```
Comment (albanD) [inline c10/util/Exception.h:238]: If we want to keep the cuda there, we could put this in a `torch::cuda::` namespace?
But on the other hand, maybe we also want our CPU allocator to raise this?

**Review by albanD — COMMENTED** [review]
Rename sounds ok even though it is a generalization that I'm not so sure about.

File: c10/util/Exception.h (hunk 2 above)
Comment (zdevito) [reply c10/util/Exception.h:238]: I think it is ok to generalize this. Previous to now it really was only used in some C++ tests to check that OOMs were being raised in certain situations. I can see a future where we extend the exception with a device attribute to describe which thing ran out of memory if we really needed the distinction.

Comment (albanD) [reply c10/util/Exception.h:238]: Sounds good!

**Review by albanD — APPROVED** [review]
LGTM

**Pull request closed without merging**